### ISSUE TITLE
Derivative of arbitrary Pytree with custom_vjp, custom_jvp

### ISSUE THREAD

**Issue by rmoyard** [pytorch/functorch#1064]
Hello Functorch!

I am working on rewriting the PennyLane pytorch interface (quantum software) and I think your project can be really useful!

Currently I am trying to bypass the limitation of `autograd.Function` that only supports returning Tensor or a tuple of Tensor as return type. I have read that you want to be able to support `autograd.Function` https://github.com/pytorch/functorch/issues/1031 in Functorch. Do you think it will be possible to define functions  (`autograd.Function`) that return arbitrarily nested tuple structure? Like in Jax I would like to define `custom_vjp` (vjp needs to match Pytree structure of the function parameters) or `custom_jvp` (jvp needs to match Pytree structure of the function return).

To summarize: when you add support for custom gradient (with autograd.Function) can you also remove the structure limitation of the function return type `autograd.Function`? Or alternatively if you don't add this support but build your own custom_vjp, custom_jvp can try to avoid any limitation for the return type of functions?

Example with `autograd.function`:
```
import functorch

class ExecuteTapesNew(torch.autograd.Function):
    @staticmethod
    def forward(ctx, param_0, param_1, param_2, param_3):
        #Support Pytree structure
        return tuple([tuple([param_0, param_1]), tuple([param_2, param_3])])

    @staticmethod
    def backward(ctx, *dys):
        jac = compute_jac(params)
        vjp = dys * jacobian
        # VJP match the structure of the parameters
        return tuple(vjp)

def execute(param_0, param_1, param_2, param_3):
    return ExecuteTapesNew.apply(param_0, param_1, param_2, param_3)

jac = functorch.jacrev(execute, argnums=(0,1,2,3))(param_0, param_1, param_2, param_3)
```
Example Jax like:
```
@jax.custom_vjp
def f(param_0, param_1, param_2, param_3):
  return tuple([tuple([param_0, param_1]), tuple([param_2, param_3])])

def f_fwd(param_0, param_1, param_2, param_3):
  return f(param_0, param_1, param_2, param_3), ()

def f_bwd(res, g):
  return (param_0*g, param_1*g, param_2*g, param_3*g)

f.defvjp(f_fwd, f_bwd)
```

**Comment by zou3519** [comment]
cc @soulitzer, @albanD

The current plan of record is that we are extending autograd.Function. As a part of this work we not planning on lifting the restriction, though your feedback is helpful and there are other things we can do.

The original reason why autograd.Function doesn't support arbitrary pytrees is that it's expensive to use pytree to peek into arbitrary python data structures. Is the additional overhead worth it to you?

In general we've been thinking of how to introduce pytrees to existing PyTorch APIs (like autograd.Function). I can see us providing some additional API to allow autograd.Function to accept pytrees. Something like a flag for autograd.Function, or giving users decorators to do so:

```
class ExecuteTapesNew(torch.autograd.Function):
    inputs_outputs_may_be_pytrees = True

    @staticmethod
    def forward(ctx, param_0, param_1, param_2, param_3):
        #Support Pytree structure
        return tuple([tuple([param_0, param_1]), tuple([param_2, param_3])])

    @staticmethod
    def backward(ctx, *dys):
        jac = compute_jac(params)
        vjp = dys * jacobian
        # VJP match the structure of the parameters
        return tuple(vjp)

```

or 

```
class ExecuteTapesNew(torch.autograd.Function):
    @staticmethod
    @pytreeify
    def forward(ctx, param_0, param_1, param_2, param_3):
        #Support Pytree structure
        return tuple([tuple([param_0, param_1]), tuple([param_2, param_3])])

    @staticmethod
    @pytreeify
    def backward(ctx, *dys):
        jac = compute_jac(params)
        vjp = dys * jacobian
        # VJP match the structure of the parameters
        return tuple(vjp)

```

**Comment by rmoyard** [comment]
Thanks for you answer @zou3519 ! After some research I am convinced that we need those nested structure, accepting a potential classical overhead. Our functions can execute multiple circuit and each of those circuit can return tuple of tuple. So here I don't see any other possibility for us.

The second solution looks very promising and we would definitely use it.

Let me know if any of these is implemented in the near future. I keep an eye on your great project.

**Comment by albanD** [comment]
Hey!
If you want to start using this today with regular PyTorch, you can already do a decorator like this.
We do want to allow more PyTorch APIs to support pytrees going forward but changing the default is hard for us. So we're considering this kind of opt-in solution. If you have time, I would be very curious to get your feedback on how well a decorator like this work for your use case and how smooth it is to use to guide further development from our end!

```python
import torch
from torch.autograd import Function
import torch.utils._pytree as pytree

# Basically wraps things in and out before passing it to the real function that the user defined.
def pytreeify(cls):
    assert issubclass(cls, Function)

    orig_fw = cls.forward
    orig_bw = cls.backward
    orig_apply = cls.apply

    def new_apply(*inp):
        flat_inp, struct = pytree.tree_flatten(inp)
        out_struct_holder = []
        flat_out = orig_apply(struct, out_struct_holder, *flat_inp)
        assert len(out_struct_holder) == 1
        return pytree.tree_unflatten(flat_out, out_struct_holder[0])

    def new_forward(ctx, struct, out_struct_holder, *flat_inp):
        inp = pytree.tree_unflatten(flat_inp, struct)
        out = orig_fw(ctx, *inp)
        flat_out, out_struct = pytree.tree_flatten(out)
        ctx._inp_struct = struct
        ctx._out_struct = out_struct
        out_struct_holder.append(out_struct)
        return tuple(flat_out)

    def new_backward(ctx, *flat_grad_outputs):
        grad_outputs = pytree.tree_unflatten(flat_grad_outputs, ctx._out_struct)
        if not isinstance(grad_outputs, tuple):
            grad_outputs = (grad_outputs,)
        grad_inputs = orig_bw(ctx, *grad_outputs)
        flat_grad_inputs, grad_inputs_struct = pytree.tree_flatten(grad_inputs)
        if grad_inputs_struct != ctx._inp_struct:
            raise RuntimeError("The backward generated an arg structure that doesn't "
                               "match the forward's input.")
        return (None, None) + tuple(flat_grad_inputs)

    cls.apply = new_apply
    cls.forward = new_forward
    cls.backward = new_backward
    return cls

# Very basic test where pytreeify is not needed
@pytreeify
class BasicTest(Function):
    @staticmethod
    def forward(ctx, inp1, inp2):
        return inp1 + inp2

    @staticmethod
    def backward(ctx, gO):
        return gO, gO

a = torch.rand(10, requires_grad=True)
b = torch.rand(10, requires_grad=True)

out = BasicTest.apply(a, b)
out.sum().backward()


# A more advanced test
@pytreeify
class Test(Function):
    @staticmethod
    def forward(ctx, tuple_holder, dict_holder):
        out0 = tuple_holder[0] + tuple_holder[1] + dict_holder["foo"]
        out1 = (dict_holder["bar"] + tuple_holder[0],)
        return (out0, out1)

    @staticmethod
    def backward(ctx, gO0, gO1):
        gtuple0 = gO0 + gO1[0]
        gtuple1 = gO0
        gdictfoo = gO0
        gdictbar = gO1[0]
        return (gtuple0, gtuple1), {"foo": gdictfoo, "bar": gdictbar}


a = torch.rand(10, requires_grad=True)
b = torch.rand(10, requires_grad=True)
c = torch.rand(10, requires_grad=True)
d = torch.rand(10, requires_grad=True)

inp1 = (a, b)
inp2 = {"foo": c, "bar": d}

out = Test.apply(inp1, inp2)
print(out)
(out[0] + out[1][0]).sum().backward()
print(a.grad)
print(b.grad)
print(c.grad)
print(d.grad)
```

**Comment by rmoyard** [comment]
Thank you so much @albanD I will give it a try and keep you posted!

**Comment by zou3519** [comment]
@rmoyard we also have support for autograd.Function with functorch transforms now in the PyTorch nightly. Guide available over at https://pytorch.org/docs/master/notes/extending.func.html .

**Comment by subho406** [comment]
As a fellow user who recently switched from Jax, this feature looks interesting and fits my use-case. Is this still being considered to be included in later PyTorch releases?

**Comment by zou3519** [comment]
> As a fellow user who recently switched from Jax, this feature looks interesting and fits my use-case. Is this still being considered to be included in later PyTorch releases?

@subho406, just to check, you're referring to the `pytreeify` decorator in albanD's comment?
